### PR TITLE
Refactor our tests descriptions to be more narrative

### LINKED ISSUE #518: Clean up our Describe/Context/It/By usage in unit tests and functional tests
- [ ] Start  using multiple Describe blocks instead of Context
- [ ] Start  using Context really as Context to Describes
- [ ] Start using the `By` function in our tests
- [ ] In general go over all our descriptions and update the descriptions to be a nice narrative

See https://onsi.github.io/ginkgo/#organizing-specs-with-containers-describe-and-context for context.

### PR DESCRIPTION
**Pull request by cynepco3hahue**
Resolves #518 

Signed-off-by: Lukianov Artyom <alukiano@redhat.com>

### CODE REVIEW

**Comment by stu-gott** [comment]
re-test this please

File: tests/vnc_test.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ import (
 	"kubevirt.io/kubevirt/tests"
 )
 
-var _ = Describe("Vmlifecycle", func() {
+var _ = Describe("VNC", func() {
```
Comment (stu-gott) [inline tests/vnc_test.go]: Good catch :)

**Comment by rmohr** [comment]
retest this please

File: tests/replicaset_test.go [hunk 2]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
 
-			Eventually(func() int32 {
-				rs, err = virtClient.ReplicaSet(tests.NamespaceTestDefault).Get(name, v12.GetOptions{})
-				Expect(err).ToNot(HaveOccurred())
-				return rs.Status.Replicas
-			}, 10, 1).Should(Equal(int32(scale)))
+		Expect(err).ToNot(HaveOccurred())
 
-			vms, err := virtClient.VM(tests.NamespaceTestDefault).List(v12.ListOptions{})
+		By("Checking a number of replicas")
```
Comment (rmohr) [inline tests/replicaset_test.go]: "Checking **the** number of relicas"

File: tests/controller_leader_election_test.go [hunk 3]
```diff
@@ -50,45 +50,50 @@ var _ = Describe("LeaderElection", func() {
 		tests.BeforeTestCleanup()
 	})
 
-	Context("Controller pod destroyed", func() {
-		It("should success to start VM", func() {
-			newLeaderPod := getNewLeaderPod(virtClient)
-			Expect(newLeaderPod).NotTo(BeNil())
+	Describe("Start a VM", func() {
+		Context("when a controller pod destroyed", func() {
```
Comment (rmohr) [inline tests/controller_leader_election_test.go]: "when the controller pod is not running"

File: tests/controller_leader_election_test.go [hunk 4]
```diff
@@ -50,45 +50,50 @@ var _ = Describe("LeaderElection", func() {
...
+				newLeaderPod := getNewLeaderPod(virtClient)
+				Expect(newLeaderPod).NotTo(BeNil())
 
-			// TODO: It can be race condition when newly deployed pod receive leadership, in this case we will need
-			// to reduce Deployment replica before destroy the pod and restore it after the test
-			Eventually(func() string {
-				leaderPodName := getLeader()
+				// TODO: It can be race condition when newly deployed pod receive leadership, in this case we will need
+				// to reduce Deployment replica before destroy the pod and restore it after the test
+				By("Destroying a controller leader pod")
```
Comment (rmohr) [inline tests/controller_leader_election_test.go]: Destroying the leading controller pod

File: tests/registry_disk_test.go [hunk 5]
```diff
@@ -68,6 +69,8 @@ var _ = Describe("RegistryDisk", func() {
 		// Verify Registry Disks are Online
 		pods, err := virtClient.CoreV1().Pods(tests.NamespaceTestDefault).List(services.UnfinishedVMPodSelector(vm))
 		Expect(err).To(BeNil())
+
+		By("Checking a number of VM disks")
```
Comment (rmohr) [inline tests/registry_disk_test.go]: Checking the number of VM disks

File: tests/registry_disk_test.go [hunk 6]
```diff
@@ -85,57 +88,73 @@ var _ = Describe("RegistryDisk", func() {
 		Expect(disksFound).To(Equal(1))
 	}
 
-	Context("Ephemeral RegistryDisk", func() {
-		It("should be able to start and stop the same VM multiple times.", func(done Done) {
-			vm := tests.NewRandomVMWithEphemeralDisk("kubevirt/cirros-registry-disk-demo:devel")
-			num := 2
-			for i := 0; i < num; i++ {
-				obj, err := virtClient.RestClient().Post().Resource("virtualmachines").Namespace(tests.NamespaceTestDefault).Body(vm).Do().Get()
+	Describe("Starting and stopping a same VM", func() {
```
Comment (rmohr) [inline tests/registry_disk_test.go]: Starting and stopping the same VM

File: tests/replicaset_test.go [hunk 7]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
+	},
+		table.Entry("to three, to two and then to zero replicas", 3, 2),
+		table.Entry("to four, to six and then to zero replicas", 5, 6),
+	)
+
+	It("should update readyReplicas once VMs are up", func() {
+		newRS := newReplicaSet()
+		doScale(newRS.ObjectMeta.Name, 2)
+
+		By("Checking a number of ready replicas")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Checking the number of ready replicas

File: tests/replicaset_test.go [hunk 8]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
-			newRS, err = virtClient.ReplicaSet(tests.NamespaceTestDefault).Create(newRS)
+	It("should remove VMs once it is marked for deletion", func() {
+		newRS := newReplicaSet()
+		// Create a replicaset with two replicas
+		doScale(newRS.ObjectMeta.Name, 2)
+		// Delete it
+		By("Deleting the VM replica set")
+		Expect(virtClient.ReplicaSet(newRS.ObjectMeta.Namespace).Delete(newRS.ObjectMeta.Name, &v12.DeleteOptions{})).To(Succeed())
+		// Wait until VMs are gone
+		By("Checking a number of VMs")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Waiting until all VMs are gone

File: tests/replicaset_test.go [hunk 9]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
-			// Create a replicaset with two replicas
-			doScale(newRS.ObjectMeta.Name, 2)
-
-			// Check for owner reference
-			vms, err := virtClient.VM(newRS.ObjectMeta.Namespace).List(v12.ListOptions{})
-			Expect(vms.Items).To(HaveLen(2))
-			for _, vm := range vms.Items {
-				Expect(vm.OwnerReferences).ToNot(BeEmpty())
+		// Delete it
+		By("Deleting the VM replica set")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Deleting the VM replica set with the "orphan" deletion strategy

File: tests/replicaset_test.go [hunk 10]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
+		orphanPolicy := v12.DeletePropagationOrphan
+		Expect(virtClient.ReplicaSet(newRS.ObjectMeta.Namespace).
+			Delete(newRS.ObjectMeta.Name, &v12.DeleteOptions{PropagationPolicy: &orphanPolicy})).To(Succeed())
+		// Wait until the replica set is deleted
```
Comment (rmohr) [inline tests/replicaset_test.go]: By("Waiting until the replica set got deleted")

File: tests/replicaset_test.go [hunk 11]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
-			}, 60*time.Second, 1*time.Second).Should(BeTrue())
-
-			vms, err = virtClient.VM(newRS.ObjectMeta.Namespace).List(v12.ListOptions{})
-			Expect(vms.Items).To(HaveLen(2))
-			for _, vm := range vms.Items {
-				Expect(vm.OwnerReferences).To(BeEmpty())
-			}
-			Expect(err).ToNot(HaveOccurred())
-		})
+		By("Checking a number of VMs")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Checking if the two VMs are orphaned and still exist

File: tests/replicaset_test.go [hunk 12]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
-			}, 10*time.Second, 1*time.Second).Should(Equal(v1.VMReplicaSetReplicaPaused))
-
-			// set new replica count while still being paused
-			rs.Spec.Replicas = tests.NewInt32(2)
-			_, err = virtClient.ReplicaSet(rs.ObjectMeta.Namespace).Update(rs)
-			Expect(err).ToNot(HaveOccurred())
+	It("should not scale when paused and scale when resume", func() {
+		rs := newReplicaSet()
+		// pause controller
+		By("Pausing")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Pausing the replicaset

File: tests/replicaset_test.go [hunk 13]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
+			rs, err = virtClient.ReplicaSet(tests.NamespaceTestDefault).Get(rs.ObjectMeta.Name, v12.GetOptions{})
 			Expect(err).ToNot(HaveOccurred())
+			if len(rs.Status.Conditions) > 0 {
+				return rs.Status.Conditions[0].Type
+			}
+			return ""
+		}, 10*time.Second, 1*time.Second).Should(Equal(v1.VMReplicaSetReplicaPaused))
+
+		// set new replica count while still being paused
+		By("Updating a number of replicas")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Updating the number of replicas

File: tests/replicaset_test.go [hunk 14]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
+		rs.Spec.Replicas = tests.NewInt32(2)
+		_, err = virtClient.ReplicaSet(rs.ObjectMeta.Namespace).Update(rs)
+		Expect(err).ToNot(HaveOccurred())
+
+		// make sure that we don't scale up
+		By("Checking a number of replicas")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Checking that the replicaset did not scale while it is paused

File: tests/replicaset_test.go [hunk 15]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
+		Consistently(func() int32 {
+			rs, err = virtClient.ReplicaSet(tests.NamespaceTestDefault).Get(rs.ObjectMeta.Name, v12.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			// Make sure that no failure happened, so that ensure that we don't scale because we are paused
+			Expect(rs.Status.Conditions).To(HaveLen(1))
+			return rs.Status.Replicas
+		}, 3*time.Second, 1*time.Second).Should(Equal(int32(0)))
+
+		// resume controller
+		By("Resuming")
```
Comment (rmohr) [inline tests/replicaset_test.go]: Resuming the controller

File: tests/replicaset_test.go [hunk 16]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
+		rs.Spec.Paused = false
+		_, err = virtClient.ReplicaSet(rs.ObjectMeta.Namespace).Update(rs)
+		Expect(err).ToNot(HaveOccurred())
+
+		// Paused condition should disappear
+		Eventually(func() int {
```
Comment (rmohr) [inline tests/replicaset_test.go]: By("Checking that the pause condition disappeared from the replicaset")

File: tests/replicaset_test.go [hunk 17]
```diff
@@ -49,161 +49,173 @@ var _ = Describe("VirtualMachineReplicaSet", func() {
...
-			// Replicas should be created
-			Eventually(func() int32 {
-				rs, err = virtClient.ReplicaSet(tests.NamespaceTestDefault).Get(rs.ObjectMeta.Name, v12.GetOptions{})
-				Expect(err).ToNot(HaveOccurred())
-				return rs.Status.Replicas
-			}, 10*time.Second, 1*time.Second).Should(Equal(int32(2)))
-		})
-
+		// Replicas should be created
+		By("Checking a number of replicas")
```
Comment (rmohr) [inline tests/replicaset_test.go]: By("Checking that the missing replicas are now created")

File: tests/vm_configuration_test.go [hunk 18]
```diff
@@ -43,36 +43,41 @@ var _ = Describe("Configurations", func() {
...
-			_, err = expecter.ExpectBatch([]expect.Batcher{
-				&expect.BExp{R: "Welcome to Alpine"},
-				&expect.BSnd{S: "\n"},
-				&expect.BExp{R: "login"},
-				&expect.BSnd{S: "root\n"},
-				&expect.BExp{R: "#"},
-				&expect.BSnd{S: "grep -c ^processor /proc/cpuinfo\n"},
-				&expect.BExp{R: "3"},
-			}, 250*time.Second)
+				By("Checking a number of CPU cores under guest OS")
```
Comment (rmohr) [inline tests/vm_configuration_test.go]: Checking the number of CPU cores under guest OS

File: tests/vm_monitoring_test.go [hunk 19]
```diff
@@ -52,40 +53,45 @@ var _ = Describe("Health Monitoring", func() {
...
-				&expect.BExp{R: "0"},
-			}, 250*time.Second)
-			Expect(err).ToNot(HaveOccurred())
-
-			namespace := vm.ObjectMeta.Namespace
-			name := vm.ObjectMeta.Name
-
-			Eventually(func() v1.VMPhase {
-				startedVM, err := virtClient.VM(namespace).Get(name, metav1.GetOptions{})
+	Describe("Watchdog device", func() {
```
Comment (rmohr) [inline tests/vm_monitoring_test.go]: A VM with a watchdog device

File: tests/vm_monitoring_test.go [hunk 20]
```diff
@@ -52,40 +53,45 @@ var _ = Describe("Health Monitoring", func() {
...
+		Context("when expires", func() {
```
Comment (rmohr) [inline tests/vm_monitoring_test.go]: You can remove the context and just write e.g.

It("should be shut down when the watchdog expires")

**Comment by cynepco3hahue** [comment]
@rmohr I fixed all comments and tried to make `By` section broader

**Comment by cynepco3hahue** [comment]
@davidvossel @rmohr Can you please review it again?

**Comment by cynepco3hahue** [comment]
ci test please

**Pull request merged**